File: SystemDesign/concepts/system-design/encryption/js/04_rsa_vs_ecc.js
/**
 * RSA vs ECC — Why modern systems prefer ECC
 *
 * RSA (1977):
 *   Security based on: difficulty of factoring large numbers (n = p × q)
 *   To double security: must roughly SQUARE the key size
 *   2048-bit RSA → 4096-bit RSA for 2x security
 *
 * ECC — Elliptic Curve Cryptography (1985, mainstream ~2010):
 *   Security based on: elliptic curve discrete logarithm problem
 *   Much harder problem → smaller keys for same security
 *
 * Key size comparison (same security level):
 *
 *   RSA key size    ECC key size    Security bits
 *   ────────────    ────────────    ─────────────
 *   1024 bit        ~160 bit        80  (broken)
 *   2048 bit        ~224 bit        112 (minimum)
 *   3072 bit        ~256 bit        128 (good)
 *   7680 bit        ~384 bit        192 (high)
 *   15360 bit       ~521 bit        256 (very high)
 *
 * ECC-256 = RSA-3072 in security, but key is 12x smaller
 *
 * Why ECC wins:
 *   ✅ Smaller keys → faster handshake, less CPU, less bandwidth
 *   ✅ Perfect Forward Secrecy (ECDHE) — new key per session
 *   ✅ Used by TLS 1.3, SSH, Bitcoin, Signal, WhatsApp
 *   ⚠️  More complex math — implementation bugs more dangerous
 *
 * Common ECC curves:
 *   P-256 (prime256v1) → NIST standard, used in TLS
 *   P-384              → higher security
 *   Curve25519         → modern, faster, used by Signal/WhatsApp
 *   secp256k1          → Bitcoin
 */

const crypto = require('crypto');

// ── Generate RSA key pair ─────────────────────────────────────────────────────
console.log('=== KEY GENERATION TIME ===\n');

console.time('RSA-2048 keygen');
const rsa = crypto.generateKeyPairSync('rsa', {
    modulusLength: 2048,
    publicKeyEncoding:  { type: 'pkcs1', format: 'pem' },
    privateKeyEncoding: { type: 'pkcs1', format: 'pem' }
});
console.timeEnd('RSA-2048 keygen');

// ── Generate ECC key pair ─────────────────────────────────────────────────────
console.time('ECC P-256 keygen');
const ecc = crypto.generateKeyPairSync('ec', {
    namedCurve: 'P-256',
    publicKeyEncoding:  { type: 'spki', format: 'pem' },
    privateKeyEncoding: { type: 'pkcs8', format: 'pem' }
});
console.timeEnd('ECC P-256 keygen');

// ── Compare key sizes ─────────────────────────────────────────────────────────
console.log('\n=== KEY SIZE COMPARISON ===');
console.log('RSA-2048 public key length:  ', rsa.publicKey.length, 'chars');
console.log('ECC P-256 public key length: ', ecc.publicKey.length, 'chars');
console.log('ECC key is', Math.round(rsa.publicKey.length / ecc.publicKey.length) + 'x smaller');

// ── Sign and verify with RSA ──────────────────────────────────────────────────
console.log('\n=== SIGN/VERIFY PERFORMANCE ===');
const message = 'Transfer $1000 to Bob — timestamp: ' + Date.now();

console.time('RSA sign');
const rsaSign = crypto.createSign('SHA256');
rsaSign.update(message);
const rsaSig = rsaSign.sign(rsa.privateKey, 'base64');
console.timeEnd('RSA sign');

console.time('RSA verify');
const rsaVerify = crypto.createVerify('SHA256');
rsaVerify.update(message);
const rsaValid = rsaVerify.verify(rsa.publicKey, rsaSig, 'base64');
console.timeEnd('RSA verify');
console.log('RSA valid:', rsaValid);

// ── Sign and verify with ECC ──────────────────────────────────────────────────
console.time('ECC sign');
const eccSign = crypto.createSign('SHA256');
eccSign.update(message);
const eccSig = eccSign.sign(ecc.privateKey, 'base64');
console.timeEnd('ECC sign');

console.time('ECC verify');
const eccVerify = crypto.createVerify('SHA256');
eccVerify.update(message);
const eccValid = eccVerify.verify(ecc.publicKey, eccSig, 'base64');
console.timeEnd('ECC verify');
console.log('ECC valid:', eccValid);

// ── ECDH — key exchange (how TLS 1.3 works) ───────────────────────────────────
console.log('\n=== ECDH KEY EXCHANGE (TLS 1.3 style) ===');
console.log('Alice and Bob each generate a key pair.');
console.log('They exchange PUBLIC keys.');
console.log('Both derive the SAME shared secret without ever sending it.');

const alice = crypto.createECDH('prime256v1');
alice.generateKeys();

const bob = crypto.createECDH('prime256v1');
bob.generateKeys();

// Alice computes shared secret using Bob's public key
const aliceShared = alice.computeSecret(bob.getPublicKey());

// Bob computes shared secret using Alice's public key
const bobShared = bob.computeSecret(alice.getPublicKey());

console.log('\nAlice shared secret:', aliceShared.toString('hex').substring(0, 20) + '...');
console.log('Bob shared secret:  ', bobShared.toString('hex').substring(0, 20) + '...');
console.log('Match?', aliceShared.toString('hex') === bobShared.toString('hex'));
console.log('\nAttacker sees both public keys but CANNOT derive the shared secret.');
console.log('This is Perfect Forward Secrecy — new keys every session.');

// ── Summary ───────────────────────────────────────────────────────────────────
console.log('\n=== RSA vs ECC SUMMARY ===');
console.log(`
Feature              RSA-2048          ECC P-256
──────────────────────────────────────────────────
Security level       ~112 bits         ~128 bits
Key size             2048 bits         256 bits
Key gen speed        slower            faster
Sign speed           slower            faster
Verify speed         faster            slower
Forward Secrecy      no (RSA-only)     yes (ECDHE)
TLS 1.3 support      limited           primary
Used in              legacy HTTPS      TLS1.3, SSH, Bitcoin
`);

File: SystemDesign/concepts/system-design/encryption/python/04_rsa_vs_ecc.py
"""
RSA vs ECC — Comparison + ECDH Key Exchange

ECC = Elliptic Curve Cryptography
  Security based on: discrete logarithm over elliptic curves
  Much harder to break than RSA for the same key size

WHY ECC WINS:
  ECC-256 bit key ≈ RSA-3072 bit key (same security, 12x smaller key)
  Faster operations (key gen, sign, verify)
  Perfect Forward Secrecy via ECDHE
  Used in: TLS 1.3, SSH, Bitcoin, Signal, WhatsApp

Install: pip install cryptography
"""

from cryptography.hazmat.primitives.asymmetric import ec
from cryptography.hazmat.primitives import hashes, serialization
from cryptography.hazmat.primitives.kdf.hkdf import HKDF
import os

# ── 1. Key Size Comparison ─────────────────────────────────────────────────────
print("=== 1. Key Size Comparison ===")
print("""
Algorithm      | Key Size  | Security Level | Notes
───────────────────────────────────────────────────────────
RSA-2048       | 2048 bits | ~112 bits      | legacy HTTPS
RSA-3072       | 3072 bits | ~128 bits      | NIST recommended
RSA-4096       | 4096 bits | ~140 bits      | large, slow
ECC P-256      |  256 bits | ~128 bits      | TLS 1.3, modern HTTPS
ECC P-384      |  384 bits | ~192 bits      | high security
ECC P-521      |  521 bits | ~260 bits      | top security
""")

# ── 2. Generate ECC Key Pair ──────────────────────────────────────────────────
print("=== 2. Generate ECC P-256 Key Pair ===")

ecc_private = ec.generate_private_key(ec.SECP256R1())   # P-256 = SECP256R1
ecc_public  = ecc_private.public_key()

# Serialize to PEM to compare key sizes
ecc_private_pem = ecc_private.private_bytes(
    encoding=serialization.Encoding.PEM,
    format=serialization.PrivateFormat.PKCS8,
    encryption_algorithm=serialization.NoEncryption()
)
ecc_public_pem = ecc_public.public_bytes(
    encoding=serialization.Encoding.PEM,
    format=serialization.PublicFormat.SubjectPublicKeyInfo
)

print(f"ECC private key PEM size: {len(ecc_private_pem)} bytes")
print(f"ECC public  key PEM size: {len(ecc_public_pem)} bytes")
print("Compare: RSA-2048 private key PEM ≈ 1700 bytes — 10x larger!")

# ── 3. ECDSA — Digital Signature with ECC ─────────────────────────────────────
print("\n=== 3. ECDSA Digital Signature ===")

document = b"Bank transfer: $1000 to Alice"

# Sign with PRIVATE key (ECDSA = Elliptic Curve Digital Signature Algorithm)
signature = ecc_private.sign(document, ec.ECDSA(hashes.SHA256()))
print(f"Signature size: {len(signature)} bytes (vs RSA signature: 256 bytes)")
print(f"Signature (hex): {signature.hex()[:40]}...")

# Verify with PUBLIC key
try:
    ecc_public.verify(signature, document, ec.ECDSA(hashes.SHA256()))
    print("Signature VALID")
except Exception:
    print("Signature INVALID")

# ── 4. ECDH — Key Exchange (the TLS 1.3 way) ─────────────────────────────────
print("\n=== 4. ECDH Key Exchange (Elliptic Curve Diffie-Hellman) ===")
print("Alice and Bob want a shared secret without sending it over the wire")
print()

# Alice generates ephemeral key pair
alice_private = ec.generate_private_key(ec.SECP256R1())
alice_public  = alice_private.public_key()
print("Alice: generated ephemeral key pair")

# Bob generates ephemeral key pair
bob_private = ec.generate_private_key(ec.SECP256R1())
bob_public  = bob_private.public_key()
print("Bob:   generated ephemeral key pair")

# They exchange PUBLIC keys (safe to send over internet)
print("\nAlice → Bob: sends alice_public")
print("Bob → Alice: sends bob_public")

# Alice computes shared secret using Bob's PUBLIC key
alice_shared = alice_private.exchange(ec.ECDH(), bob_public)

# Bob computes shared secret using Alice's PUBLIC key
bob_shared = bob_private.exchange(ec.ECDH(), alice_public)

print(f"\nAlice's shared secret: {alice_shared.hex()[:30]}...")
print(f"Bob's   shared secret: {bob_shared.hex()[:30]}...")
print(f"Secrets match? {alice_shared == bob_shared}")

# Derive AES key from shared secret
def derive_aes_key(shared_secret: bytes) -> bytes:
    hkdf = HKDF(algorithm=hashes.SHA256(), length=32, salt=None, info=b"ecdh key")
    return hkdf.derive(shared_secret)

alice_aes_key = derive_aes_key(alice_shared)
bob_aes_key   = derive_aes_key(bob_shared)
print(f"\nAES key match? {alice_aes_key == bob_aes_key}")
print("Both Alice and Bob now have the SAME AES key — attacker cannot derive it!")

# ── 5. Forward Secrecy Explained ─────────────────────────────────────────────
print("\n=== 5. Why ECDHE gives Forward Secrecy ===")
print("""
RSA TLS (no forward secrecy):
  Server has ONE long-term private key.
  If attacker records all traffic today, steals key in 5 years →
  can decrypt ALL past sessions!

ECDHE TLS (perfect forward secrecy):
  Each session generates NEW ephemeral key pair (that's the 'E' in ECDHE).
  Session key is derived, used, then DISCARDED.
  If server's long-term key is stolen → past sessions STILL safe.
  Attacker only compromises future sessions until key is rotated.

TLS 1.3 requires forward secrecy — only ECDHE and DHE allowed.
TLS 1.2 made it optional — many servers still use RSA key exchange.
""")
